src/cloud-sql-instance.ts

```diff
@@ -98,60 +107,153 @@ export class CloudSQLInstance {
   async forceRefresh(): Promise<void> {
     // if a refresh is already ongoing, just await for its promise to fulfill
     // so that a new instance info is available before reconnecting
-    if (this.ongoingRefreshPromise) {
-      await this.ongoingRefreshPromise;
+    if (this.next) {
+      await this.next;
       return;
     }
     this.cancelRefresh();
     return this.refresh();
   }
 
   async refresh(): Promise<void> {
+    const currentRefreshId = this.scheduledRefreshID;
+
     // Since forceRefresh might be invoked during an ongoing refresh
     // we keep track of the ongoing promise in order to be able to await
     // for it in the forceRefresh method.
     // In case the throttle mechanism is already initialized, we add the
     // extra wait time `limitRateInterval` in order to limit the rate of
     // requests to Cloud SQL Admin APIs.
-    this.ongoingRefreshPromise = this.throttle
-      ? this.throttle(this._refresh).call(this)
-      : this._refresh();
-
-    // awaits for the ongoing promise to resolve, since the refresh is
-    // completed once the promise is resolved, we just free up the reference
-    // to the promise at this point, ensuring any new call to `forceRefresh`
-    // is able to trigger a new refresh
-    await this.ongoingRefreshPromise;
-    this.ongoingRefreshPromise = undefined;
-
-    // Initializing the rate limiter at the end of the function so that the
-    // first refresh cycle is never rate-limited, ensuring there are 2 calls
-    // allowed prior to waiting a throttle interval.
+    this.next = (
+      this.throttle && this.scheduledRefreshID
+        ? this.throttle(this.performRefresh).call(this)
+        : this.performRefresh()
+    )
+      // These needs to be part of the chain of promise referenced in
+      // next in order to avoid race conditions
+      .then((nextValues: RefreshResult) => {
+        // in case the id at the moment of starting this refresh cycle has
+        // changed, that means that it has been canceled
+        if (currentRefreshId !== this.scheduledRefreshID) {
+          return;
+        }
+
+        // In case the performRefresh method succeeded
+        // then we go ahead and update values
+        this.updateValues(nextValues);
+
+        this.scheduleRefresh();
+
+        // This is the end of the successful refresh chain, so now
+        // we release the reference to the next
+        this.next = undefined;
+      })
+      .catch((err: unknown) => {
+        // In case there's already an active connection we won't throw
+        // refresh errors to the final user, scheduling a new refresh instead.
+        if (this.stablishedConnection) {
+          if (currentRefreshId === this.scheduledRefreshID) {
+            this.scheduleRefresh();
```

Comment:
What happens to caller here? Do we block them until a new refresh is done? What happens if the refresh continues to fail? Do we block the event loop?

Reply:
No, we never block the event loop. What happens is that the current metadata value is still used for connections in that case.